tycho-indexer/src/services/rpc.rs

```diff
@@ -718,6 +732,144 @@ where
+    ) -> Result<dto::TracedEntryPointRequestResponse, RpcError> {
+        let pagination_params: PaginationParams = (&request.pagination).into();
+
+        // For consistency, we pre-apply pagination to ensure it's predetermined which components
+        // will appear on which page.
+        // Apply pagination to the contract addresses. This is done so that we can determine which
+        // contracts were not returned from the db and get them from the buffer instead.
+        let mut paginated_component_ids: Option<Vec<String>> = None;
+        if let Some(component_ids) = request.component_ids {
+            paginated_component_ids = Some(
+                component_ids
+                    .iter()
+                    .skip(pagination_params.offset() as usize)
+                    .take(pagination_params.page_size as usize)
+                    .cloned()
+                    .collect(),
+            );
+        }
+
+        let filter = EntryPointFilter {
+            protocol_system: request.protocol_system,
+            component_ids: paginated_component_ids,
+        };
+
+        let entry_points_tracing_params_data = self
+            .db_gateway
+            .get_entry_points_tracing_params(filter, Some(&pagination_params))
+            .await
+            .map_err(|err| {
+                error!(error = %err, "Error while getting entry points with tracing params.");
+                err
+            })?;
+
+        trace!(
+            entry_points_tracing_params = ?entry_points_tracing_params_data,
+            "Retrieved entry points with tracing params from database."
+        );
+
+        // Flatten the ID lists, throwing away component ids, to avoid making duplicate db calls
+        // when getting traced entry points.
+        let entry_point_ids: HashSet<EntryPointId> = entry_points_tracing_params_data
+            .entity
+            .values()
+            .flat_map(|entry_points_with_tracing_params| {
+                entry_points_with_tracing_params
+                    .iter()
+                    .map(|entry_point| {
+                        entry_point
+                            .entry_point
+                            .external_id
+                            .clone()
+                    })
+            })
+            .collect();
+
+        let traced_entry_points = self
+            .db_gateway
+            .get_traced_entry_points(&entry_point_ids)
+            .await
+            .map_err(|err| {
+                error!(error = %err, "Error while getting traced entry points.");
+                err
+            })?;
+
+        trace!(
+            traced_entry_points = ?traced_entry_points,
+            "Retrieved traced entry points entry points from database."
+        );
+
+        let mut traced_entry_points_by_component = HashMap::new();
+
+        for (component_id, entry_points_set) in entry_points_tracing_params_data.entity {
+            let mut pairs = Vec::new();
```

Comment:
Small optimisation suggestion: if you know the length of a Vec before you populate it, you can already set the Vec capacity when you instantiate it. This allows the compiler to know the Vec size from the start and it doesn't need to reallocate the vec as it grows. 
```suggestion
            let mut pairs = Vec::with_capacity(entry_points_set.len());
```
Optimisation isn't crazy important here so this small improvement is just me finding a teaching moment and isn't actually necessary.

Reply:
Oh, cool! Good to know!

Reply:
You could actually still use it here. The capacity can be a best-guess, does not need to be 100% accurate. And the likelihood of there being a failed trace is hopefully low enough that your best guess is accurate.